packages/filecoin-address/src/index.ts

```diff
@@ -171,38 +185,28 @@ export function decode(address: string): Address {
 }
 
 export function encode(coinType: string, address: Address): string {
-  if (!address || !address.str) throw Error('Invalid address')
+  if (!address || !address.bytes) throw Error('Invalid address')
+
+  const protocol = address.protocol()
   const payload = address.payload()
+  const prefix = `${coinType}${protocol}`
 
-  switch (address.protocol()) {
-    case 0: {
-      return (
-        coinType +
-        String(address.protocol()) +
-        leb.unsigned.decode(address.payload())
-      )
+  switch (protocol) {
+    case Protocol.ID: {
+      return `${prefix}${leb.unsigned.decode(payload)}`
     }
     case Protocol.DELEGATED: {
-      const protocolByte = new Uint8Array([address.protocol()])
-      const namespace = payload[0]
-      const subAddr = payload.slice(1)
-      const checksum = getChecksum(uint8arrays.concat([protocolByte, payload]))
-
-      return (
-        String(coinType) +
-        String(address.protocol()) +
-        String(namespace) +
-        'f' +
-        base32.encode(uint8arrays.concat([subAddr, checksum]))
-      )
+      const checksum = getChecksum(address.bytes)
+      const nsBytes = new Uint8Array(payload, 0, namespaceByteLen)
+      const namespace = new Int64(nsBytes).toNumber()
+      const subAddr = payload.slice(namespaceByteLen)
+      const bytes = uint8arrays.concat([subAddr, checksum])
+      return `${prefix}${namespace}f${base32.encode(bytes)}`
```

Comment:
This makes sure we always strip the right amount of bytes from the payload for the namespace